src/ray/core_worker/core_worker.cc

```diff
@@ -1783,31 +1794,51 @@ Status CoreWorker::GetLocationFromOwner(
   }
 
+  absl::flat_hash_map<rpc::Address, std::vector<ObjectID>> objects_by_owner;
+  for (const auto &object_id : object_ids) {
+    rpc::Address owner_address;
+    RAY_RETURN_NOT_OK(GetOwnerAddress(object_id, &owner_address));
+    objects_by_owner[owner_address].push_back(object_id);
+  }
+
   auto mutex = std::make_shared<absl::Mutex>();
-  auto num_remaining = std::make_shared<size_t>(object_ids.size());
+  auto num_remaining = std::make_shared<size_t>(objects_by_owner.size());
   auto ready_promise = std::make_shared<std::promise<void>>();
   auto location_by_id =
       std::make_shared<absl::flat_hash_map<ObjectID, std::shared_ptr<ObjectLocation>>>();
 
-  for (const auto &object_id : object_ids) {
-    rpc::Address owner_address;
-    RAY_RETURN_NOT_OK(GetOwnerAddress(object_id, &owner_address));
+  for (const auto &pair : objects_by_owner) {
+    const auto &owner_address = pair.first;
+    const auto &owner_object_ids = pair.second;
     auto client = core_worker_client_pool_->GetOrConnect(owner_address);
     rpc::GetObjectLocationsOwnerRequest request;
-    auto object_location_request = request.mutable_object_location_request();
-    object_location_request->set_intended_worker_id(owner_address.worker_id());
-    object_location_request->set_object_id(object_id.Binary());
+    request.set_intended_worker_id(owner_address.worker_id());
+    // add multi object address and id into one request
+    for (const auto &object_id : owner_object_ids) {
+      request.add_object_id(object_id.Binary());  // Add each object ID to the request
+    }
+
     client->GetObjectLocationsOwner(
         request,
-        [object_id, mutex, num_remaining, ready_promise, location_by_id](
-            const Status &status, const rpc::GetObjectLocationsOwnerReply &reply) {
+        [owner_object_ids,
+         mutex,
+         num_remaining,
+         ready_promise,
+         location_by_id,
+         owner_address](const Status &status,
+                        const rpc::GetObjectLocationsOwnerReply &reply) {
           absl::MutexLock lock(mutex.get());
           if (status.ok()) {
-            location_by_id->emplace(
-                object_id, std::make_shared<ObjectLocation>(CreateObjectLocation(reply)));
+            auto locations = CreateObjectLocations(reply);
+            for (size_t i = 0; i < owner_object_ids.size(); ++i) {
+              // Assuming successful query, emplace the result
+              location_by_id->emplace(owner_object_ids[i],
+                                      std::make_shared<ObjectLocation>(locations[i]));
+            }
           } else {
-            RAY_LOG(WARNING) << "Failed to query location information for " << object_id
-                             << " with error: " << status.ToString();
+            RAY_LOG(WARNING)
+                << "Failed to query location information for objects owned by "
+                << owner_address.worker_id() << " with error: " << status.ToString();
```

Comment:
<< "Failed to query location information for objects " << debug_string(owner_object_ids) << " owned by " << ....